docs/source/en/kv_cache.md

```diff
@@ -0,0 +1,222 @@
+    <p>Therefore, when iteratively calling `forward()` instead of the `generate()` method, it’s crucial to ensure that the attention mask shape matches the combined length of past and current key-values. The attention mask should have the shape `(batch_size, past_kv_length + new_tokens_length)`. This is usually handled internally when you call `generate()` method. If you want to implement your own generation loop with Cache classes, take this into consideration and prepare the attention mask to hold values to current and past tokens.</p>
+    <p>See an example below for how to implement your own generation loop.</p>
+    <pre><code class="language-python">
+    import torch
+    from transformers import AutoTokenizer, AutoModelForCausalLM, DynamicCache
+    model_id = "meta-llama/Llama-2-7b-chat-hf"
+    model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16, device_map='auto')
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+    past_key_values = DynamicCache()
+    messages = [{"role": "user", "content": "Hello, what's your name."}]
+    inputs = tokenizer.apply_chat_template(messages, add_generation_prompt=True, return_tensors="pt", return_dict=True).to(model.device)
+    generated_ids = inputs.input_ids
+    max_new_tokens = 10
+    for _ in range(max_new_tokens):
+        outputs = model(**inputs, past_key_values=past_key_values, use_cache=True)     
+        # Greedily sample one next token
+        next_token_ids = outputs.logits[:, -1:].argmax(-1)
+        generated_ids = torch.cat([generated_ids, next_token_ids], dim=-1)   
+        # Prepare inputs for the next generation step by leaaving unprocessed tokens, in our case we have only one new token
+        # and expanding attn mask for the new token, as explained above
+        attention_mask = torch.cat(inputs.attention_mask)
+        attention_mask = torch.cat([attention_mask, attention_mask.new_ones((attention_mask.shape[0], 1))], dim=-1)
+        inputs = {"input_ids": next_token_ids, "attention_mask": attention_mask}
+    print(tokenizer.batch_decode(generated_ids, skip_special_tokens=True))
+    </code></pre>
+</details>
+
+
+
+## Generate with Cache
+
+In 🤗 Transformers, we support various Cache types to optimize the performance across different models and tasks. By default all models generate with caching
+and for that we use the ["DynamicCache"] class. It allows us to dynamically grow cache size, by saving more and more keys and values as we generate. Usually all you have to do to enable or disable it is to pass in `use_cache=True` or `use_cache=False` into the `generate()` method.
+
+Apart from that we support many cache classes, which can be set with a `cache_implementation` argument. It can be set to one of `["static", "quantized", "sliding_window", "hybrid"]`. Let's look into each cache type in more details.
```

Comment:
Instead of enumerating specific values for `cache_implementation` here, let's link to the field in the documentation (which should have the full list always up to date!)